### PR TITLE
put marker on root element

### LINKED ISSUE #3335: Warn on bad runtime version
E.g. I saw this when I tried to deploy a `<runtime>java9</runtime>` app:

image

(And we really should do a better job of displaying these error messages)

### LINKED ISSUE #3340: Java 7 quickfix warning even if <runtime>java8 present
Not sure how I got into this. Was applying the quickfix and doing Ctrl+z / Ctrl+Shift+z. Note that the _Problems_ view does not have the corresponding warning entry.

selection_002

selection_001

### PR DESCRIPTION
**Pull request by elharo**
fix #3340 
fix #3335 
@chanseokoh

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java [hunk 1]
```diff
@@ -107,21 +107,26 @@ public void testValidate_dynamicWebProject() throws CoreException, ValidationExc
...
         + "<runtime>java8</runtime>"
-        + "</appengine-web-app>").getBytes(StandardCharsets.UTF_8);
-    validator.validate(reporter, null, xml);
+        + "</appengine-web-app>";
+    
+    IProject project = appEngineStandardProject.getProject();
+    IFile file = project.getFile("testdata.xml");
+    file.create(ValidationTestUtils.stringToInputStream(APPLICATION_XML), 0, null);
+    
+    validator.validate(reporter, file, xml.getBytes(StandardCharsets.UTF_8));
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java]: `file` is `testdata.xml`, whose content is `APPLICATION_XML`. Also, the `xml.getBytes()` content is different from `APPLICATION_XML`. It is not clear if this validates the content of `testdata.xml` or `xml`. Maybe the code should be `file.create(stringToInputStream(xml), 0, null)`?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java [hunk 2]
```diff
@@ -107,21 +107,26 @@ public void testValidate_dynamicWebProject() throws CoreException, ValidationExc
...
     List<IMessage> messages = reporter.getMessages();
     if (!messages.isEmpty()) {
       Assert.fail(messages.get(0).getText());
     }
   }
 
   @Test
-  public void testValidate() throws IOException {
+  public void testValidate() throws IOException, CoreException {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java]: `CoreException` probably not needed.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidator.java [hunk 3]
```diff
@@ -84,8 +84,10 @@ public void validate(IValidationContext helper, IReporter reporter) throws Valid
   /**
    * Adds an {@link IMessage} to the XML file for every
    * {@link ElementProblem} found in the file.
+   * @throws CoreException 
    */
-  void validate(IReporter reporter, IFile source, byte[] bytes) throws IOException {
+  @VisibleForTesting
+  void validate(IReporter reporter, IFile source, byte[] bytes) throws IOException, CoreException {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidator.java]: `CoreException` should be removed, probably. The Javadoc too.

**Comment by chanseokoh** [comment]
> fix #3335

Does it fix #3335? Wrong issue number?

**Comment by chanseokoh** [comment]
Just realized Ctrl+1 is effective only at the start of the content. As you said, we should fix #3363.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java (hunk 1 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java (hunk 2 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidatorTest.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidator.java (hunk 3 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XmlSourceValidator.java]: done

**Pull request merged**